File: apps/api/app/schemas/edition.py
import uuid
from datetime import date, datetime

from pydantic import BaseModel, field_validator

from app.models.enums import EditionStatus, EditionType


class EditionCreate(BaseModel):
    number: int | None = None
    year: int
    type: EditionType = EditionType.NORMAL
    title: str | None = None
    subtitle: str | None = None
    publication_date: date

    @field_validator("number")
    @classmethod
    def number_positive(cls, v: int | None) -> int | None:
        if v is not None and v < 1:
            raise ValueError("Number must be >= 1")
        return v

    @field_validator("year")
    @classmethod
    def year_valid(cls, v: int) -> int:
        if v < 2000 or v > 2100:
            raise ValueError("Year out of range")
        return v


class EditionUpdate(BaseModel):
    title: str | None = None
    subtitle: str | None = None
    publication_date: date | None = None


class NextEditionNumberResponse(BaseModel):
    year: int
    type: EditionType
    next_number: int
    auto_numbering: bool


class EditionItemOut(BaseModel):
    id: uuid.UUID
    matter_id: uuid.UUID
    matter_title: str = ""
    section_title: str | None = None
    position: int
    page_number: int | None = None

    model_config = {"from_attributes": True}


class EditionResponse(BaseModel):
    id: uuid.UUID
    number: int
    year: int
    type: EditionType
    title: str
    subtitle: str | None
    publication_date: date
    status: EditionStatus
    created_by: uuid.UUID
    published_at: datetime | None
    created_at: datetime
    updated_at: datetime
    items: list[EditionItemOut] = []
    item_count: int = 0

    model_config = {"from_attributes": True}


class EditionListResponse(BaseModel):
    id: uuid.UUID
    number: int
    year: int
    type: EditionType
    title: str
    status: EditionStatus
    publication_date: date
    created_at: datetime
    item_count: int = 0
    signature_count: int = 0

    model_config = {"from_attributes": True}


class AddItemRequest(BaseModel):
    matter_id: uuid.UUID
    section_title: str | None = None
    position: int | None = None


class ReorderItem(BaseModel):
    id: uuid.UUID
    position: int


class ReorderRequest(BaseModel):
    items: list[ReorderItem]


class SignRequest(BaseModel):
    signing_credential_id: uuid.UUID | None = None
    pfx_base64: str | None = None
    pfx_password: str | None = None
    reason: str = "Assinatura de Edição"
    location: str = ""
    visible: bool = False


class SignResponse(BaseModel):
    verification_code: str
    signed_pdf_hash: str
    certificate_subject: str
    certificate_serial: str
    signed_at: str
    message: str = "Edition signed successfully"


class ValidateSignatureResponse(BaseModel):
    edition_id: str
    status: str
    signed_at: str
    certificate_subject: str
    certificate_serial: str
    certificate_thumbprint: str
    verification_code: str
    issues: list[str] = []
    recommendation: str = "OK"


class PublishResponse(BaseModel):
    edition_id: str
    status: str
    published_at: str
    verification_code: str
    immutability_hash: str
    message: str

File: apps/web-admin/src/types/edition.ts
export type EditionStatus = "draft" | "reviewing" | "scheduled" | "closed" | "pdf_generated" | "signed" | "published" | "cancelled";
export type EditionType = "normal" | "extra" | "suplementar";

export interface EditionItem {
  id: string;
  matter_id: string;
  matter_title: string;
  section_title: string | null;
  position: number;
  page_number: number | null;
}

export interface Edition {
  id: string;
  number: number;
  year: number;
  type: EditionType;
  title: string;
  subtitle: string | null;
  publication_date: string;
  status: EditionStatus;
  created_by: string;
  published_at: string | null;
  created_at: string;
  updated_at: string;
  items: EditionItem[];
  item_count: number;
}

export interface EditionListItem {
  id: string;
  number: number;
  year: number;
  type: EditionType;
  title: string;
  status: EditionStatus;
  publication_date: string;
  created_at: string;
  item_count: number;
  signature_count: number;
}

export interface MatterListItem {
  id: string;
  title: string;
  status: string;
  act_type_id: string;
  summary: string | null;
  created_at: string;
}
